### ISSUE TITLE
micropip: support extras

### ISSUE THREAD

**Issue by rth** [pyodide/pyodide#2535]
As reported by @Zac-HD in https://github.com/pyscript/pyscript/issues/105 , currently micropip ignores extras in requirements for instance,
```py
>>> import micropip
>>> await micropip.install("hypothesis[cli]")
```
doesn't install black while it's in the `cli` extras.

Related: 
> [pip.pypa.io/en/stable/cli/pip_install/#requirement-specifiers](https://pip.pypa.io/en/stable/cli/pip_install/#requirement-specifiers) and [peps.python.org/pep-0508/#extras](https://peps.python.org/pep-0508/#extras)

**Comment by amitsaha** [comment]
I will have a go at this one.

**Comment by hoodmane** [comment]
We already try to support extras but apparently not successfully:
https://github.com/pyodide/pyodide/blob/b0f7feb210060b21da35b2d6bf047d362d1790b5/packages/micropip/src/micropip/_micropip.py#L392

**Comment by ryanking13** [comment]
> We already try to support extras but apparently not successfully:
> 
> https://github.com/pyodide/pyodide/blob/b0f7feb210060b21da35b2d6bf047d362d1790b5/packages/micropip/src/micropip/_micropip.py#L392

Yeah... it is apparently a bug. Thanks for working on this @amitsaha!

**Comment by amitsaha** [comment]
The bug is I believe here: https://github.com/pyodide/pyodide/blob/b0f7feb210060b21da35b2d6bf047d362d1790b5/packages/micropip/src/micropip/_micropip.py#L324

This condition here, evaluates to `False` since the `extra` "env" is set to an empty string by default. I am not exactly sure what the right fix is here. One guess and a "fix" I am contemplating is this change:

```
 if req.marker:
+            if str(req.marker._markers[0][0]) == "extra":
+                self.ctx["extra"] = str(req.marker._markers[0][2])
```
The above change will result in setting the value of `extra` to whatever the value of the `extra` requires the package requirement describes. I have to read up more on the spec and perhaps the implementation of `pip` to better understand whether this is the right fix or not. Any thoughts, please share.

**Comment by amitsaha** [comment]
Thanks @ryanking13 - appreciate the insights. For now, based on the `pip` snippet, i have this code: https://github.com/pyodide/pyodide/commit/d51eaeeb6ab28e029efcc0eead7176b5fe76f5a2 

Some notes:

`marker` is set only on the requirement(s) which are pulled in due to the `extra` marker on the original requirement. That is, for `hypothesis[cli]`, `req.marker` evaluates to `True` only for say, `black` which is pulled due to the `cli` extra. 

Additionally, the `extras` attribute is only set on the `hypothesis` requirement object. So, what i am doing now, is setting up `ctx_extras` with the `extras` when it finds it, and then similar to `pip` evaluates the environment based on *any* of the `extra` values found in this list. 

I am going to work on a better unit test and then create a PR.

**Comment by amitsaha** [comment]
PR is ready for review

**Comment by amitsaha** [comment]
Shall we close this issue @hoodmane ?

**Comment by hoodmane** [comment]
Closed by #2584. Thanks again @amitsaha.
